File: quant_web/src/api/system.ts
// quant_web/src/api/system.ts
import request from '@/utils/request'
import { handleResponse } from '@/utils/responseHandler'
import {
  ApiResponse,
  PaginatedResponse,
  HealthCheckResponse
} from '@/types/api'
import {SystemLog} from "@/types/entities";

/**
 * 系统管理API服务
 * 提供系统状态监控、配置管理、日志查询和系统维护功能
 */

export interface SystemStatus {
  version: string;
  uptime: number;
  status: 'running' | 'stopped' | 'error';
  last_startup: string;
  engine_status: {
    main_engine: string;
    event_engine: string;
    strategy_engine: string;
    data_service: string;
  };
}

export interface SystemSettings {
  data_sync: {
    auto_sync: boolean;
    sync_interval: number;
    data_sources: string[];
  };
  trading: {
    commission_rate: number;
    tax_rate: number;
    min_commission: number;
  };
  risk: {
    max_position_ratio: number;
    max_daily_loss: number;
    enable_auto_stop: boolean;
  };
  notification: {
    enable_email: boolean;
    enable_wechat: boolean;
    risk_notification: boolean;
  };
}

export interface ConnectionStatus {
  database: boolean;
  redis: boolean;
  tushare: boolean;
  broker: boolean;
  last_check: string;
}

export interface ResourceUsage {
  cpu_percent: number;
  memory_percent: number;
  memory_used: number;
  memory_total: number;
  disk_usage: number;
  network_io: {
    bytes_sent: number;
    bytes_recv: number;
  };
}

export interface DatabaseStatus {
  total_tables: number;
  total_records: number;
  stock_data_count: number;
  trade_data_count: number;
  last_optimized: string;
}

export interface SystemLogQueryParams {
  level?: 'INFO' | 'WARNING' | 'ERROR' | 'DEBUG';
  module?: string;
  page?: number;
  limit?: number;
  start_time?: string;
  end_time?: string;
}

export default {
  /**
   * 获取系统状态
   * @returns 系统状态信息
   */
  async getSystemStatus(): Promise<SystemStatus> {
    return request.get('/system/status')
      .then(handleResponse)
      .then((data: ApiResponse<SystemStatus>) => data.data)
  },

  /**
   * 获取系统日志
   * @param params 查询参数
   * @returns 系统日志分页结果
   */
  async getSystemLogs(params?: SystemLogQueryParams): Promise<PaginatedResponse<SystemLog>> {
    return request.get('/system/logs', { params })
      .then(handleResponse)
      .then((data: PaginatedResponse<SystemLog>) => data)
  },

  /**
   * 触发数据同步
   * @param params 同步参数
   * @returns 同步任务信息
   */
  async triggerDataSync(params: {
    data_type: string;
    start_date?: string;
    end_date?: string;
    symbols?: string[];
  }): Promise<{ task_id: string }> {
    return request.post('/system/data/sync', params)
      .then(handleResponse)
      .then((data: ApiResponse<{ task_id: string }>) => data.data)
  },

  /**
   * 获取数据同步状态
   * @returns 数据同步状态信息
   */
  async getDataSyncStatus(): Promise<any> {
    return request.get('/system/data/status')
      .then(handleResponse)
      .then((data: ApiResponse<any>) => data.data)
  },

  /**
   * 获取系统配置
   * @returns 系统配置信息
   */
  async getSystemSettings(): Promise<SystemSettings> {
    return request.get('/system/settings')
      .then(handleResponse)
      .then((data: ApiResponse<SystemSettings>) => data.data)
  },

  /**
   * 更新系统配置
   * @param settings 配置更新参数
   * @returns 更新后的系统配置
   */
  async updateSystemSettings(settings: Partial<SystemSettings>): Promise<SystemSettings> {
    return request.put('/system/settings', settings)
      .then(handleResponse)
      .then((data: ApiResponse<SystemSettings>) => data.data)
  },

  /**
   * 获取连接状态
   * @returns 各服务连接状态
   */
  async getConnections(): Promise<ConnectionStatus> {
    return request.get('/system/connections')
      .then(handleResponse)
      .then((data: ApiResponse<ConnectionStatus>) => data.data)
  },

  /**
   * 获取资源使用情况
   * @returns 系统资源使用信息
   */
  async getResources(): Promise<ResourceUsage> {
    return request.get('/system/resources')
      .then(handleResponse)
      .then((data: ApiResponse<ResourceUsage>) => data.data)
  },

  /**
   * 获取数据库状态
   * @returns 数据库状态信息
   */
  async getDatabaseStatus(): Promise<DatabaseStatus> {
    return request.get('/system/database')
      .then(handleResponse)
      .then((data: ApiResponse<DatabaseStatus>) => data.data)
  },

  /**
   * 健康检查
   * @returns 健康检查结果
   */
  async healthCheck(): Promise<HealthCheckResponse> {
    return request.get('/system/health')
      .then(handleResponse)
      .then((data: HealthCheckResponse) => data)
  },

  /**
   * 清理系统缓存
   * @returns 清理操作结果
   */
  async clearCache(): Promise<{ cleared: boolean; message: string }> {
    return request.post('/system/cache/clear')
      .then(handleResponse)
      .then((data: ApiResponse<{ cleared: boolean; message: string }>) => data.data)
  },

  /**
   * 重启系统服务
   * @param service 服务名称
   * @returns 重启操作结果
   */
  async restartService(service: string): Promise<{ success: boolean; message: string }> {
    return request.post(`/system/services/${service}/restart`)
      .then(handleResponse)
      .then((data: ApiResponse<{ success: boolean; message: string }>) => data.data)
  }
}

File: quant_server/api/system.py
# quant_server/api/system.py
from datetime import datetime

from fastapi import APIRouter, Depends, Query, Body
from typing import Optional

from quant_server.api.dependencies import get_data_service
from quant_server.db.data_service import DataService

router = APIRouter(prefix="/api/system", tags=["system"])

@router.get("/status")
async def get_system_status(data_service: DataService = Depends(get_data_service)):
    """获取系统状态"""
    # 这里需要实现系统状态获取逻辑
    # 暂时返回模拟数据
    return {
        "status": "running",
        "data_connections": {
            "tushare": "connected",
            "database": "connected"
        },
        "last_data_sync": "2023-08-23T08:00:00",
        "uptime": "5 days, 12 hours"
    }

@router.get("/logs")
async def get_system_logs(
    level: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    data_service: DataService = Depends(get_data_service)
):
    """获取系统日志"""
    # 这里需要实现系统日志获取逻辑
    # 暂时返回模拟数据
    return {
        "data": [
            {
                "timestamp": "2023-08-23T10:30:00",
                "level": "INFO",
                "module": "data_sync",
                "message": "Daily data sync completed"
            }
        ],
        "pagination": {
            "page": page,
            "limit": limit,
            "total": 1,
            "pages": 1
        }
    }

@router.post("/data/sync")
async def trigger_data_sync(
    sync_params: dict = Body(...),
    data_service: DataService = Depends(get_data_service)
):
    """触发数据同步"""
    # 这里需要实现数据同步逻辑
    # 暂时返回模拟数据
    return {
        "sync_id": f"sync_{datetime.now().timestamp()}",
        "status": "started",
        "params": sync_params
    }

@router.get("/data/status")
async def get_data_sync_status(data_service: DataService = Depends(get_data_service)):
    """获取数据同步状态"""
    # 这里需要实现数据同步状态获取逻辑
    # 暂时返回模拟数据
    return {
        "last_sync": "2023-08-23T08:00:00",
        "status": "completed",
        "records_updated": 12500
    }

@router.get("/settings")
async def get_system_settings(data_service: DataService = Depends(get_data_service)):
    """获取系统设置"""
    # 这里需要实现系统设置获取逻辑
    # 暂时返回模拟数据
    return {
        "data_source": "tushare",
        "auto_sync": True,
        "sync_time": "08:00",
        "risk_limits": {
            "max_position_per_stock": 0.2,
            "max_daily_loss": -0.05
        }
    }

@router.put("/settings")
async def update_system_settings(
    settings: dict = Body(...),
    data_service: DataService = Depends(get_data_service)
):
    """更新系统设置"""
    # 这里需要实现系统设置更新逻辑
    # 暂时返回模拟数据
    return {"updated": True, "settings": settings}


@router.get("/connections")
async def get_connections(data_service: DataService = Depends(get_data_service)):
    """获取系统连接状态"""
    return { "dataSource": True, "tradeGateway": True, "strategyEngine": True }
@router.get("/resources")
async def get_resources(data_service: DataService = Depends(get_data_service)):
    """获取系统资源使用情况"""
    return { "cpu": 0.45, "memory": 0.62, "disk": 0.28, "network": 0.15 }

@router.get("/database")
async def get_database(data_service: DataService = Depends(get_data_service)):
    """获取数据库状态"""
    return { "size": "1024MB", "tables": 42, "lastBackup": "2023-08-23T08:00:00" }